### PR TITLE
Provide for not tracking certain internal memory descriptor types.

### PR DESCRIPTION
**Pull request by gbtitus**
Add a "track this" indicator to the table of memory descriptors, along
with a function to return that indicator and code to pay attention to it
in the memory tracking implementation.  This will allow us to not track
certain runtime internal allocations that don't have a well-defined
connection to constructs in Chapel code, such as internal descriptors
for Active Memory handling.

For now all such indicators are actually set so that everything is still
tracked.  In the future I will change some of them to turn tracking off,
in order to help resolve https://chapel.atlassian.net/browse/CHAPEL-8.

The modification with the largest impact in this commit is actually not
a functional one.  In order to reformat the CHPL_MD_ALL_MEMDESCS() macro
definition to make it still be readable with the addition of the "track"
indicator, I shortened a number of the memory descriptor enum names and
some of the corresponding strings.  This part of the commit is wholly
responsible for the following files changing:
  runtime/src/chpl-privatization.c
  runtime/src/chpl-string-support.c
  runtime/src/chpl-string.c
  runtime/src/chplsys.c
  runtime/src/comm/gasnet/comm-gasnet.c
  runtime/src/comm/none/comm-none.c
  runtime/src/config.c
  runtime/src/tasks/fifo/tasks-fifo.c
  runtime/src/threads/pthreads/threads-pthreads.c

The other files changed as follows.

runtime/include/chpl-mem-desc.h
runtime/src/chpl-mem-desc.c
  Add the "track" indicator to CHPL_MD_ALL_MEMDESCS() and adjust uses of
  this macro to take that indicator into account.  Make some associated
  internal changes to improve readability.  Add chpl_mem_descTrack(), to
  return the "track" indicator.
  And of course, while here, change a number of the enum names and a few
  of the corresponding description strings, as described above.

runtime/src/chplmemtrack.c
  Track or don't track specific memory descriptor types depending on the
  setting of the "track" indicator for that type and the value of the
  overriding CHPL_RT_MEMTRACK_ALL_MDS environment variable.

runtime/include/chpl-env.h (new)
runtime/src/chpl-env.c (new)
  Add chpl_get_rt_env() and chpl_get_rt_env_bool().  The latter gets the
  presumed boolean value of a specified CHPL_RT_* environment variable,
  with a default.  It uses the former, which gets the string value of a
  specified CHPL_RT_* environment variable, with a default.  (Thes new
  functions can also now be reused in other places in the runtime where
  we need the values of environment variables.)

### CODE REVIEW

**Comment by gbtitus** [comment]
@hildeth , @ronawho , this is a replacement for PR #1911. I did it this way because after Elliot's initial reviewer of #1911 I moved the new get-environment-variable function out of chplsys.[hc] and into a new chpl-env.[hc], and I wanted to leave chplsys.[hc] without lingering changes due to the addition and then removal of that function. I've left #1911 for reference, but I'll delete it after this one is merged.

**Comment by ronawho** [comment]
:+1:

File: runtime/include/chpl-mem-desc.h [hunk 1]
```diff
@@ -39,110 +41,57 @@
...
+  m(THREAD_LIST_DESC,     "thread list descriptor",                    true), \
+  m(IO_BUFFER,            "io buffer or bytes",                        true), \
+  m(OS_LAYER_TMP_DATA,    "OS layer temporary data",                   true), \
+  m(GMP,                  "gmp data",                                  true), \
+  m(GETS_PUTS_STRIDES,    "put_strd/get_strd array of strides",        true), \
+  m(NUM,                  "*** this must be the last entry ***",       true)
 
 
 //
 // Define the numeration constants for the memory descriptors.
```
Comment (hildeth) [inline runtime/include/chpl-mem-desc.h]: Typo: "numeration" --> "enumeration".
Maybe next time....

**Comment by hildeth** [comment]
This looks fine.
In some cases, it might be nice to track a small subset of memory descriptors rather than just all or none. Having an immediate need for that feature will probably drive its implementation, but if you have ideas for the design, you could add that to comments to guide that effort.
It looks like user-defined memory descriptors are always tracked, which is probably the most useful setting.

File: runtime/include/chpl-mem-desc.h (hunk 1 above)
Comment (gbtitus) [inline runtime/include/chpl-mem-desc.h]: Oopsie! Will fix ...

**Comment by gbtitus** [comment]
Yeah, I'll add a comment to that effect.

**Comment by hildeth** [comment]
The typo is long-standing: not your fault.
The design notes will be good.  Thanks.
THH
________________________________

**Comment by gbtitus** [comment]
I've made one final change, moving the tracking control entirely into chpl-mem-desc and documenting that we could do finer control in the future. I'm going to do a bit more testing and then merge.

**Pull request merged**